File: apps/java/src/main/java/com/functions/DnsResolveFunction.java
package com.functions;

import com.microsoft.azure.functions.*;
import com.microsoft.azure.functions.annotation.*;
import java.net.InetAddress;
import java.util.Optional;

/**
 * DNS resolution probe — resolves hostname for network troubleshooting.
 */
public class DnsResolveFunction {

    @FunctionName("dnsResolve")
    public HttpResponseMessage run(
            @HttpTrigger(
                name = "req",
                methods = {HttpMethod.GET},
                authLevel = AuthorizationLevel.ANONYMOUS,
                route = "dns/{hostname=google.com}")
            HttpRequestMessage<Optional<String>> request,
            @BindingName("hostname") String hostname,
            final ExecutionContext context) {

        context.getLogger().info("DNS resolve request for: " + hostname);

        try {
            InetAddress[] addresses = InetAddress.getAllByName(hostname);
            StringBuilder ips = new StringBuilder("[");
            for (int i = 0; i < addresses.length; i++) {
                if (i > 0) ips.append(",");
                ips.append("\"").append(addresses[i].getHostAddress()).append("\"");
            }
            ips.append("]");

            String body = String.format(
                "{\"hostname\":\"%s\",\"addresses\":%s,\"resolved\":true}",
                hostname, ips.toString());

            return request.createResponseBuilder(HttpStatus.OK)
                    .header("Content-Type", "application/json")
                    .body(body)
                    .build();
        } catch (Exception e) {
            String body = String.format(
                "{\"hostname\":\"%s\",\"error\":\"%s\",\"resolved\":false}",
                hostname, e.getMessage());

            return request.createResponseBuilder(HttpStatus.OK)
                    .header("Content-Type", "application/json")
                    .body(body)
                    .build();
        }
    }
}

File: apps/dotnet/Functions/DnsResolveFunction.cs
using System.Net;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.Functions.Worker;
using Microsoft.Extensions.Logging;

namespace AzureFunctionsGuide.Functions;

/// <summary>
/// DNS resolution probe — resolves a hostname and returns addresses.
/// Route: GET /api/dns/{hostname}
/// </summary>
public class DnsResolveFunction
{
    private readonly ILogger<DnsResolveFunction> _logger;

    public DnsResolveFunction(ILogger<DnsResolveFunction> logger)
    {
        _logger = logger;
    }

    [Function("dnsResolve")]
    public async Task<IActionResult> Run(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "dns/{hostname}")] HttpRequest req,
        string hostname)
    {
        _logger.LogInformation("DNS resolve requested for {Hostname}", hostname);

        try
        {
            var addresses = await Dns.GetHostAddressesAsync(hostname);
            var addressList = addresses.Select(a => a.ToString()).ToArray();

            _logger.LogInformation("DNS resolved {Hostname} to {Addresses}", hostname, string.Join(", ", addressList));

            return new OkObjectResult(new
            {
                hostname,
                addresses = addressList
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "DNS resolution failed for {Hostname}", hostname);

            return new ObjectResult(new
            {
                hostname,
                error = ex.Message
            })
            { StatusCode = 502 };
        }
    }
}
